### PR TITLE
[SPMD] Mesh to support custom device order.

### PR DESCRIPTION
**Pull request by yeounoh**
This implements `Mesh` class from #3871 , to support custom device order in logical XLA device mesh topology.

### CODE REVIEW

File: test/test_xla_sharding.py [hunk 1]
```diff
@@ -11,38 +11,60 @@
...
 from torch_xla.experimental.xla_sharded_tensor import XLAShardedTensor
+from torch_xla.experimental.pjrt import using_pjrt
 
 
-@unittest.skipIf((os.getenv('PJRT_DEVICE') == "") or (
-    xm.get_xla_supported_devices("GPU") is not None
-), "PyTorch/XLA SPMD requires PJRT_DEVICE={CPU, TPU}, GPU is currently not supported."
-                )
+@unittest.skipIf(not using_pjrt() or xm.get_xla_supported_devices("GPU"),
+                 f"Requires PJRT_DEVICE set to `TPU` or `CPU`.")
```
Comment (JackCaoG) [inline test/test_xla_sharding.py:18]: @will-cromar I think PJRT-GPU single core is ready now?

File: torch_xla/csrc/tensor_util.cpp [hunk 2]
```diff
@@ -925,7 +925,7 @@ std::vector<torch::lazy::BackendDataPtr> CreateTensorsData(
       // host communications. This means that we may need to manually shard
       // across global devices for multi-host training.
       std::vector<std::string> local_devices =
-          xla::ComputationClient::Get()->GetLocalDevices();
+          xla::ComputationClient::Get()->GetAllDevices();
```
Comment (JackCaoG) [inline torch_xla/csrc/tensor_util.cpp:928]: why this change?

File: torch_xla/experimental/xla_sharding.py [hunk 3]
```diff
@@ -1,14 +1,74 @@
...
 
 
-def mark_sharding(t: Union[torch.Tensor,
-                           XLAShardedTensor], mesh_shape: Tuple[int],
+class Mesh:
+  """Describe the logical XLA device topology mesh and the underlying resources.
+
+  Args:
+    device_ids (Union[np.ndarray, List]): A raveled list of devices (IDs) in a custom order. The list is reshaped
+        to an `mesh_shape` array, filling the elements using C-like index order. For example,
```
Comment (JackCaoG) [inline torch_xla/experimental/xla_sharding.py:17]: where is the example lol?

Comment (JackCaoG) [reply torch_xla/experimental/xla_sharding.py:17]: oh ok it is below, you might want to change the wording here.

File: torch_xla/csrc/tensor_util.cpp (hunk 2 above)
Comment (yeounoh) [reply torch_xla/csrc/tensor_util.cpp:928]: It's the same on the single host, but on the pod -- I need to list all the devices available.

File: test/test_xla_sharding.py (hunk 1 above)
Comment (yeounoh) [reply test/test_xla_sharding.py:18]: It's blocked from the our SPMD side, once we support TPU, the transition should be easier to GPU -- maybe sometime next year once we are done with the basic/core SPMD features?

File: torch_xla/experimental/xla_sharding.py (hunk 3 above)
Comment (yeounoh) [reply torch_xla/experimental/xla_sharding.py]: Done

File: test/test_xla_sharding.py [hunk 4]
```diff
@@ -11,38 +11,60 @@
...
-    xst1 = xs.mark_sharding(xt1, mesh_shape, partition_spec)
+    xst1 = xs.mark_sharding(xt1, self._get_mesh((1, self.n_devices)),
+                            partition_spec)
 
     # TODO(244003536) add more tests for XLAShardedTensror.
     self.assertTrue(isinstance(xst1, XLAShardedTensor))
 
+  def test_custom_tile_assignment(self):
+    xt = torch.randn(10, 20).to(device=xm.xla_device())
+    mesh_shape = (1, self.n_devices)
```
Comment (jonb377) [inline test/test_xla_sharding.py:48]: I see the tests have all devices mapped to a single axis - is there anything stopping us from using e.g. `mesh_shape = (2, self.n_devices / 2)`?

**Review by JackCaoG — APPROVED** [review]
Thanks!

File: torch_xla/experimental/xla_sharding.py [hunk 5]
```diff
@@ -1,14 +1,74 @@
...
+  """
+
+  device_ids: np.ndarray
+  mesh_shape: Tuple[int, ...]
+  axis_names: Tuple[str, ...]
+
+  def __init__(self,
+               device_ids: Union[np.ndarray, List],
+               mesh_shape: Tuple[int, ...],
+               axis_names: Tuple[str, ...] = None):
```
Comment (jonb377) [inline torch_xla/experimental/xla_sharding.py:48]: Just curious - how will `axis_names` be used long-term? Is it just for annotating the mesh?

**Review by jonb377 — APPROVED** [review]
LGTM, thanks!

File: torch_xla/experimental/xla_sharding.py (hunk 5 above)
Comment (yeounoh) [reply torch_xla/experimental/xla_sharding.py:48]: Good question, mesh axis annotation is useful since it makes the annotation logic more readable. We can also build a partitioning rule based on the axis name, instead of int indices.

File: test/test_xla_sharding.py (hunk 4 above)
Comment (yeounoh) [reply test/test_xla_sharding.py:48]: No, but for the unit testing a flat mesh is easier to work with since we don't know how many devices we would have (e.g., for CPU, we will have 1).

File: torch_xla/experimental/xla_sharding.py [hunk 6]
```diff
@@ -18,32 +78,36 @@ def mark_sharding(t: Union[torch.Tensor,
         mesh_shape (Tuple[Union[int, None]]): A int tuple describing the logical topology
         of the device mesh, and each element describes the number of devices in
         the corresponding axis.
```
[the remark this answers is not in the document]
Comment (yeounoh) [reply torch_xla/experimental/xla_sharding.py]: Good catch :)

**Comment by yeounoh** [comment]
`narrow_copy_dense` has been renamed to `narrow_copy_dense_symint` in the upstream, rebasing to fix the build issue.

**Review by steventk-g — APPROVED** [review]
lgtm!
